internal-services/catalog/pulp-push-disk-images-task.yaml

```diff
@@ -0,0 +1,191 @@
+        echo "$pulpKey" > "$PULP_KEY"
+        echo "$udcacheCert" > "$UDCACHE_CERT"
+        echo "$udcacheKey" > "$UDCACHE_KEY"
+        set -x
+
+        DISK_IMAGE_DIR="$(mktemp -d)"
+
+        process_component() { # Expected argument is [component json]
+            COMPONENT=$1
+            PULLSPEC=$(jq -r '.containerImage' <<< "${COMPONENT}")
+            DESTINATION="${DISK_IMAGE_DIR}/$(jq -r '.staged.destination' <<< "${COMPONENT}")/FILES"
+            mkdir -p "${DESTINATION}"
+            DOWNLOAD_DIR=$(mktemp -d)
+            cd $DOWNLOAD_DIR
+            # oras has very limited support for selecting the right auth entry,
+            # so create a custom auth file with just one entry
+            AUTH_FILE=$(mktemp)
+            select-oci-auth "${PULLSPEC}" > "$AUTH_FILE"
+            oras pull --registry-config "$AUTH_FILE" "$PULLSPEC"
+            NUM_MAPPED_FILES=$(jq '.staged.files | length' <<< ${COMPONENT})
+            for ((i = 0; i < $NUM_MAPPED_FILES; i++)) ; do
+                FILE=$(jq -c --arg i "$i" '.staged.files[$i|tonumber]' <<< $COMPONENT)
+                SOURCE=$(jq -r '.source' <<< $FILE)
+                FILENAME=$(jq -r '.filename' <<< $FILE)
+                gzip -d "${SOURCE}.gz" || echo "didn't find mapped file: ${SOURCE}.gz"
+                DESTINATION_FILE="${DESTINATION}/${FILENAME}"
+                # Albeit a very small one, this is a race condition since this is run in parallel
+                if [ -f "${DESTINATION_FILE}" ] ; then
+                    echo -n "Multiple files use the same destination value: $DESTINATION"
+                    echo " and filename value: $FILENAME. Failing..."
+                    exit 1
+                fi
+                mv $SOURCE "${DESTINATION_FILE}"
+            done
+        }
+
+        N=$(params.concurrentLimit)  # The maximum number of images to be pulled at once
+        RUNNING_JOBS="\j" # Bash parameter for number of jobs currently running
+        NUM_COMPONENTS=$(jq '.components | length' <<< "$SNAPSHOT_JSON")
+        # Process each component in parallel
+        for ((i = 0; i < $NUM_COMPONENTS; i++)) ; do
+            COMPONENT=$(jq -c --arg i "$i" '.components[$i|tonumber]' <<< "$SNAPSHOT_JSON")
+            # Limit batch size to concurrent limit
+            while (( ${RUNNING_JOBS@P} >= N )); do
+                wait -n
+            done
+            process_component $COMPONENT &
+        done
+
+        # Wait for remaining processes to finish
+        while (( ${RUNNING_JOBS@P} > 0 )); do
+            wait -n
+        done
+
+        # Change to the subdir with the images
+        cd ${DISK_IMAGE_DIR}
+
+        STAGED_JSON='{"header":{"version": "0.2"},"payload":{"files":[]}}'
+
+        # Add the files to the payload
+        for file in $(find * -type f); do
+            STAGED_JSON=$(jq --arg filename $(basename $file) --arg path "$file" \
+              '.payload.files[.payload.files | length] += 
```

Comment:
It's a small detail, but I think the `+` in `+=` is redundant here, because you're assigning this to a new array index and it didn't exist previously, so there is nothing really to add to.

Another option would be to use `+=` on the array level and wrap the new item in `[]` - this way you don't need to specify the new array index:
```sh
STAGED_JSON=$(jq --arg filename $(basename $file) --arg path "$file" \
    '.payload.files += 
    [{"filename": $filename, "relative_path": $path}]' <<< $STAGED_JSON)
```

Reply:
doesn't seem to work
```
$ STAGED_JSON='{"header":{"version": "0.2"},"payload":{"files":[]}}'
$ for file in $(find * -type f) ; do   STAGED_JSON=$(jq --arg filename $(basename $file) --arg path "$file" \
    '.payload.files[.payload.files | length] = [{"filename": $filename, "relative_path": $path}]' <<< $STAGED_JSON); done
$ echo $STAGED_JSON | yq -P -I4
header:
    version: "0.2"
payload:
    files:
        - - filename: a
            relative_path: 1/FILES/a
        - - filename: b
            relative_path: 1/FILES/b
        - - filename: a
            relative_path: 2/FILES/a
        - - filename: b
            relative_path: 2/FILES/b
```

```
$ STAGED_JSON='{"header":{"version": "0.2"},"payload":{"files":[]}}'
$ for file in $(find * -type f) ; do   STAGED_JSON=$(jq --arg filename $(basename $file) --arg path "$file" \
    '.payload.files[.payload.files | length] += {"filename": $filename, "relative_path": $path}' <<< $STAGED_JSON); done
$ echo $STAGED_JSON | yq -P -I4
header:
    version: "0.2"
payload:
    files:
        - filename: a
          relative_path: 1/FILES/a
        - filename: b
          relative_path: 1/FILES/b
        - filename: a
          relative_path: 2/FILES/a
        - filename: b
          relative_path: 2/FILES/b
```

Reply:
Changed to your first suggestion